script/lib/task.js

```diff
@@ -0,0 +1,138 @@
+    /** A human readable name for what this Task does. */
+    this.name = name;
+
+    /** A prefix to apply to all print operations. Adjusted appropriately when this is called by a parent Task. */
+    this.indent = "";
+  }
+
+  /** Indicate that execution may continue if this Task throws an error. */
+  canFail() {
+    return false;
+  }
+
+  /** Indicate that this Task should be skipped. Passed the same arguments as `run` would be. Returning a string will print that string as the reason for skipping. May be a value instead. Must be synchronous. */
+  skip() {
+    return false;
+  }
+
+  /** The body of execution for the Task. Override this when defining a new Task. Do not call directly, use `start` instead. Arguments may be passed via the `start` method. */
+  run() {
+    throw new Error("Task must override `run` method");
+  }
+
+  /** The way to make a Task start running. Arguments passed here will be given to the `run` method. The return value of the `run` method will be wrapped in a promise and returned. */
+  async start(...args) {
+    let skip = this.skip;
+    if (typeof skip === "function") {
+      skip = skip(...args);
+    }
+
+    if (skip) {
+      if (typeof skip === "string") {
+        this.print(`Skipping ${this.name}: ${skip}`, "?");
+      } else {
+        this.print(`Skipping ${this.name}`, "?");
+      }
+      return;
+    }
+
+    this.print(`Running ${this.name}`, ">");
+
+    try {
+      return this.run(...args);
+    } catch (e) {
+      if ((typeof this.canFail === "function" && this.canFail(e)) || this.canFail) {
+        this.warn(`Task failed: ${e.msg}`);
+        return;
+      }
+      throw e;
+    }
+  }
+
+  /** Used to launch a Task inside this Task's `run` method. The first argument is the task to launch, the others and return value are the same as `start`. */
+  async subtask(task, ...args) {
+    const oldIndent = task.indent;
+    task.indent = this.indent + "  ";
+    const result = task.start(...args);
+    result.finally(() => task.indent = oldIndent);
+    return result;
+  }
+
+  /** Log information that is gratuitous. */
+  silly(msg) {
+    this.print(msg, "  :");
+  }
+
+  /** Log information that is possibly useful to debug the task, but not needed on normal runs. */
+  verbose(msg) {
+    this.print(msg, "  *");
+  }
+
+  /** Log information that is useful to see during execution. */
+  info(msg) {
+    this.print(msg, "  -");
+  }
+
+  /** Indicate that a recoverable problem occurred. E.g., an expected file was missing, but might not be needed anyways. */
+  warn(msg) {
+    this.print(msg, "!");
+  }
+
+  /** Indicate that an unrecoverable problem occurred. See `canFail` to indicate if this should propagate to the parent Task (as an exception), or silently resolve this Task. */
+  error(msg) {
+    this.print(msg, "  !!!");
+  }
+
+  /** Log the current work being done by this Task. Typically for work that is not a Task itself. */
+  update(msg) {
+    this.print(msg, "  >>");
+  }
```

Comment:
It would be good to use Azure syntax for some of these
https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#formatting-commands

Reply:
Should be fine. The current values are more like placeholders, just to get the pretty output. Looks like sub tasks could use the `##[group]` syntax too.